sphinx/config.py

```diff
@@ -159,30 +148,30 @@ class Config(object):
                                 'env'),
     )  # type: Dict[unicode, Tuple]
 
-    def __init__(self, dirname, filename, overrides, tags):
-        # type: (unicode, unicode, Dict, Tags) -> None
+    def __init__(self, *args):
+        # type: (Any) -> None
+        if len(args) == 4:
+            # old style arguments: (dirname, filename, overrides, tags)
+            warnings.warn('The argument of Config() class has been changed. '
+                          'Use Config.read() to read configuration from conf.py.',
+                          RemovedInSphinx30Warning)
+            dirname, filename, overrides, tags = args
+            if dirname is None:
```

Comment:
I still don't think this should go here. Instead, it should go into `sphinx.application.Application`. I've added a comment to this effect above.

Reply:
Ohhh, wait - this is only for the old style config. Yup, this is fine so.